src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -3838,11 +3838,29 @@ private static bool CanReuseParameter(CSharp.Syntax.ParameterSyntax parameter)
                 type = null;
                 name = this.EatToken(SyntaxKind.ArgListKeyword);
             }
-            var exclamation = this.CurrentToken.Kind == SyntaxKind.ExclamationToken ? this.EatToken(SyntaxKind.ExclamationToken) : null;
+            SyntaxToken exclamation = null;
+            SyntaxToken equals = null;
+            if (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
+            {
+                exclamation = this.EatToken(SyntaxKind.ExclamationToken);
+            }
+            else if (this.CurrentToken.Kind == SyntaxKind.ExclamationEqualsToken)
+            {
+                var notEq = this.EatToken();
+                exclamation = new SyntaxToken(SyntaxKind.ExclamationToken);
+                equals = new SyntaxToken(SyntaxKind.EqualsToken);
+                notEq = AddLeadingSkippedSyntax(notEq, SyntaxFactory.MissingToken(SyntaxKind.WhitespaceTrivia));
+                exclamation = this.AddError(exclamation, ErrorCode.ERR_NeedSpaceBetweenExclamationAndEquals);
+
+            }
+            //var exclamation = this.CurrentToken.Kind == SyntaxKind.ExclamationToken ? this.EatToken(SyntaxKind.ExclamationToken) : null;
             EqualsValueClauseSyntax def = null;
             if (this.CurrentToken.Kind == SyntaxKind.EqualsToken)
             {
-                var equals = this.EatToken(SyntaxKind.EqualsToken);
+                equals = this.EatToken(SyntaxKind.EqualsToken);
+            }
+            if (!(equals is null))
+            {
                 var value = this.ParseExpressionCore();
                 def = _syntaxFactory.EqualsValueClause(equals, value: value);
```

Comment:
having multiple tokens exist for the same spans violates how roslyn does tree invariants.  If you see `!=` in this situation, and you're trying to recover gracefully, you must do the following:

1.  Create two seperate tokens for the `!` and the `=`.  The `!` token must have the leading trivia from teh `!=` token, and the `=` token must have the trailing trivia of the `!=` token.  
2. You should only do this work if the `!=` does not already have any diagnostics on it (otherwise you'll lose them).
3. from this point on, you must not use the `!=` token anymore, and you should instead only use the `!` and `=` tokens.

You should also be extremely careful and separate out all these cases o that there's no weird behavior going on here.  as it stands, it's hard to track the above to ensure that the right thing is happening.

Reply:
3 is already done; will focus on 1 and 2

Reply:
You say "3 is already done" . But in the code, i see you doing stuff like doing this:

>  `notEq = AddLeadingSkippedSyntax(notEq, SyntaxFactory.MissingToken(SyntaxKind.WhitespaceTrivia))`

You really cannot use notEq from this point on, it will just never be correct.

Reply:
Ik i'm getting rid of it, changes are just local.

Reply:
We don't seem to be splitting the token anymore, so much of that code is deleted anyway.